In [1]:
import pandas as pd
import numpy as np
import os
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [2]:
train_df = pd.read_csv('MB train.csv')
train_df.head()


ID       y  X0 X1  X2 X3 X4 X5 X6 X8  X10  X11  X12  X13  X14  X15  X16  \
0   0  130.81   k  v  at  a  d  u  j  o    0    0    0    1    0    0    0   
1   6   88.53   k  t  av  e  d  y  l  o    0    0    0    0    0    0    0   
2   7   76.26  az  w   n  c  d  x  j  x    0    0    0    0    0    0    0   
3   9   80.62  az  t   n  f  d  x  l  e    0    0    0    0    0    0    0   
4  13   78.02  az  v   n  f  d  h  d  n    0    0    0    0    0    0    0   

   X17  X18  X19  X20  X21  X22  X23  X24  X26  X27  X28  X29  X30  X31  X32  \
0    0    1    0    0    1    0    0    0    0    0    0    0    0    1    0   
1    0    1    0    0    0    0    0    0    0    1    0    0    0    1    0   
2    1    0    0    0    0    0    0    0    0    1    1    1    0    1    0   
3    0    0    0    0    0    0    0    0    0    1    1    1    0    1    0   
4    0    0    0    0    0    0    0    0    0    1    1    1    0    1    0   

   X33  X34  X35  X36  X37  X38  X39  X40  X41  X42  X43  X44  X45  X46  X47  \
0    0    0    1    0    1    0    0    0    0    0    0    0    0    1    0   
1    0    0    1    0    1    0    0    0    0    0    0    0    0    0    0   
2    0    0    1    0    1    0    0    0    0    0    1    0    0    1    0   
3    0    0    1    0    1    0    0    0    0    0    1    0    0    1    0   
4    0    0    1    0    1    0    0    0    0    0    1    0    0    1    0   

   X48  X49  X50  X51  X52  X53  X54  X55  X56  X57  X58  X59  X60  X61  X62  \
0    0    0    0    0    0    0    0    0    0    0    1    0    0    0    0   
1    0    0    0    1    0    0    0    0    0    0    0    0    0    1    0   
2    0    0    0    1    0    0    1    0    0    0    1    0    0    1    0   
3    0    0    0    0    0    0    1    0    0    0    0    0    0    1    0   
4    0    0    0    1    0    0    1    0    0    0    0    0    0    1    0   

   X63  X64  X65  X66  X67  X68  X69  X70  X71  X73  X74  X75  X76  X77  X78  \
0    0    0    0    0    0    1    0    1    0    0    1    0    0    0    0   
1    0    0    0    0    0    0    0    1    0    0    1    0    0    0    0   
2    0    0    0    0    0    1    0    1    1    0    1    1    1    0    0   
3    0    0    0    0    0    0    0    1    1    0    1    0    1    0    0   
4    0    0    0    0    0    0    0    1    0    0    1    0    1    0    0   

   X79  X80  X81  X82  X83  X84  X85  X86  X87  X88  X89  X90  X91  X92  X93  \
0    0    0    0    0    0    0    1    0    0    0    0    0    0    0    0   
1    0    1    0    0    0    0    1    0    0    0    0    0    0    0    0   
2    0    1    0    0    0    1    1    0    0    0    0    0    0    0    0   
3    0    1    0    0    0    1    0    0    0    0    0    0    0    0    0   
4    0    1    0    0    0    0    0    0    0    0    0    0    0    0    0   

   X94  X95  X96  X97  X98  X99  X100  X101  X102  X103  X104  X105  X106  \
0    0    0    0    0    0    0     0     0     0     0     0     0     0   
1    0    0    1    0    1    0     1     1     0     0     0     0     0   
2    0    0    1    0    1    0     0     1     0     0     0     0     0   
3    0    0    1    0    1    0     0     1     0     0     0     0     0   
4    0    0    1    0    1    0     0     1     0     0     0     0     0   

   X107  X108  X109  X110  X111  X112  X113  X114  X115  X116  X117  X118  \
0     0     0     0     0     1     0     0     1     0     1     0     1   
1     0     0     0     0     1     0     0     0     0     0     0     1   
2     0     0     0     0     1     0     0     0     0     0     0     0   
3     0     1     0     0     1     0     0     1     0     0     0     0   
4     0     1     0     0     1     0     0     1     0     0     0     0   

   X119  X120  X122  X123  X124  X125  X126  X127  X128  X129  X130  X131  \
0     1     1     0     0     0     0     0     0     1     0     0     1   
1     1     1     0     0     0     0     0     1

In [3]:
test_df = pd.read_csv('MB test.csv')
test_df.head()


ID  X0 X1  X2 X3 X4 X5 X6 X8  X10  X11  X12  X13  X14  X15  X16  X17  X18  \
0   1  az  v   n  f  d  t  a  w    0    0    0    0    0    0    0    0    0   
1   2   t  b  ai  a  d  b  g  y    0    0    0    0    0    0    0    0    0   
2   3  az  v  as  f  d  a  j  j    0    0    0    0    1    0    0    0    0   
3   4  az  l   n  f  d  z  l  n    0    0    0    0    0    0    0    0    0   
4   5   w  s  as  c  d  y  i  m    0    0    0    0    1    0    0    0    0   

   X19  X20  X21  X22  X23  X24  X26  X27  X28  X29  X30  X31  X32  X33  X34  \
0    0    0    0    0    0    0    0    1    1    1    0    1    0    0    0   
1    1    0    0    0    0    0    0    1    0    0    0    1    0    0    0   
2    0    0    0    0    0    0    0    1    0    1    0    1    0    0    0   
3    0    0    0    0    0    0    0    1    1    1    0    1    0    0    0   
4    0    0    0    0    0    0    0    1    0    0    0    0    0    0    0   

   X35  X36  X37  X38  X39  X40  X41  X42  X43  X44  X45  X46  X47  X48  X49  \
0    1    0    1    0    0    0    0    0    1    0    0    1    0    0    0   
1    1    0    1    0    0    0    0    0    0    0    1    1    0    0    1   
2    1    0    1    0    0    0    0    0    0    0    0    0    0    0    0   
3    1    0    1    0    0    0    0    0    1    0    0    1    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   X50  X51  X52  X53  X54  X55  X56  X57  X58  X59  X60  X61  X62  X63  X64  \
0    0    0    0    0    1    0    0    0    0    0    0    1    0    0    0   
1    1    1    0    0    0    0    0    0    1    0    0    1    0    0    0   
2    0    1    0    0    1    0    0    0    1    0    0    1    0    0    0   
3    0    0    0    0    1    0    0    0    0    0    0    1    0    0    0   
4    0    1    0    0    0    0    0    0    1    0    0    1    0    0    1   

   X65  X66  X67  X68  X69  X70  X71  X73  X74  X75  X76  X77  X78  X79  X80  \
0    0    0    0    0    0    1    0    0    1    0    1    0    0    0    1   
1    0    0    0    0    0    1    0    0    1    0    0    0    0    0    1   
2    0    0    0    0    0    1    0    0    1    0    1    0    0    0    1   
3    0    0    0    0    0    1    0    0    1    0    1    0    0    0    1   
4    0    0    0    0    0    1    0    0    1    0    0    0    0    0    1   

   X81  X82  X83  X84  X85  X86  X87  X88  X89  X90  X91  X92  X93  X94  X95  \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    1    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    1    0    0    0    0    0    0    0    0    0    0   

   X96  X97  X98  X99  X100  X101  X102  X103  X104  X105  X106  X107  X108  \
0    1    0    1    0     0     1     0     0     0     0     0     0     0   
1    0    0    1    0     0     1     0     1     0     0     0     0     0   
2    1    0    1    0     1     1     0     1     0     0     0     0     0   
3    1    0    1    0     0     1     0     0     0     0     0     0     0   
4    1    0    1    0     1     1     0     1     0     0     1     0     0   

   X109  X110  X111  X112  X113  X114  X115  X116  X117  X118  X119  X120  \
0     0     0     1     0     0     1     0     0     0     0     0     0   
1     0     0     1     0     0     0     0     1     0     1     1     1   
2     0     0     1     0     0     0     0     0     0     0     0     1   
3     0     0     1     0     0     1     0     0     0     0     0     1   
4     0     0     1     0     0     0     0     0     0     1     1     1   

   X122  X123  X124  X125  X126  X127  X128  X129  X130  X131  X132  X133  \
0     0     0     0     0     0     0     1     0     0     0     1     0   
1     0     0     0     0

In [4]:
train_df.shape

(4209, 378)

In [5]:
test_df.shape

(4209, 377)

In [6]:
#combine test and trin datasets in one datasets
train_df['source'] = 'train_df'
test_df['source'] = 'test_df'
data = pd.concat([train_df , test_df], ignore_index = True)
print(data.shape)

(8418, 379)


In [7]:
pd.options.display.float_format = '{:.4f}'.format
variance = data.var()
variance = variance.reset_index()
variance.columns = ["ID","values"]
variance= variance.sort_values("values",ascending=1)
variance

C:\Users\devan\AppData\Local\Temp\ipykernel_30136\3210523354.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  variance = data.var()


ID       values
219  X233       0.0001
315  X330       0.0001
276  X290       0.0001
279  X293       0.0001
221  X235       0.0001
283  X297       0.0001
3     X11       0.0001
254  X268       0.0001
243  X257       0.0001
282  X296       0.0001
281  X295       0.0001
354  X369       0.0002
177  X190       0.0002
83    X93       0.0002
275  X289       0.0002
324  X339       0.0002
245  X259       0.0002
332  X347       0.0002
274  X288       0.0002
196  X210       0.0002
266  X280       0.0002
222  X236       0.0004
357  X372       0.0004
113  X124       0.0004
264  X278       0.0004
97   X107       0.0005
304  X319       0.0005
24    X33       0.0005
30    X39       0.0005
246  X260       0.0005
85    X95       0.0005
256  X270       0.0005
190  X204       0.0005
191  X205       0.0005
238  X252       0.0005
368  X384       0.0006
7     X15       0.0006
255  X269       0.0006
317  X332       0.0007
33    X42       0.0008
193  X207       0.0008
303  X318       0.0008
231  X245       0.0008
50    X59       0.0009
77    X87       0.0009
141  X153       0.0009
134  X145       0.0011
367  X383       0.0011
172  X184       0.0011
248  X262       0.0011
148  X160       0.0011
155  X167       0.0011
73    X83       0.0012
252  X266       0.0012
100  X110       0.0012
244  X258       0.0012
31    X40       0.0013
79    X89       0.0013
95   X105       0.0013
76    X86       0.0014
263  X277       0.0014
351  X366       0.0014
64    X74       0.0014
369  X385       0.0015
239  X253       0.0015
51    X60       0.0015
82    X92       0.0017
257  X271       0.0018
342  X357       0.0018
234  X248       0.0018
58    X67       0.0020
199  X213       0.0020
179  X192       0.0020
185  X199       0.0020
102  X112       0.0020
56    X65       0.0020
81    X91       0.0021
16    X24       0.0021
13    X21       0.0021
94   X104       0.0023
338  X353       0.0024
8     X16       0.0026
292  X307       0.0027
87    X97       0.0028
320  X335       0.0028
112  X123       0.0030
268  X282       0.0031
171  X183       0.0031
350  X365       0.0031
267  X281       0.0031
226  X240       0.0032
349  X364       0.0032
295  X310       0.0034
27    X36       0.0037
285  X299       0.0038
284  X298       0.0038
297  X312       0.0038
114  X125       0.0039
213  X227       0.0039
153  X165       0.0040
25    X34       0.0045
240  X254       0.0045
216  X230       0.0045
21    X30       0.0050
310  X325       0.0056
46    X55       0.0058
160  X172       0.0059
53    X62       0.0059
202  X216       0.0059
223  X237       0.0060
355  X370       0.0061
198  X212       0.0063
44    X53       0.0064
225  X239       0.0064
200  X214       0.0064
92   X102       0.0064
68    X78       0.0065
302  X317       0.0066
186  X200       0.0066
294  X309       0.0068
157  X169       0.0070
17    X26       0.0070
323  X338       0.0073
305  X320       0.0074
84    X94       0.0075
78    X88       0.0077
229  X243       0.0077
111  X122       0.0077
228  X242       0.0078
80    X90       0.0078
365  X380       0.0080
203  X217       0.0081
207  X221       0.0081
366  X382       0.0081
9     X17       0.0081
293  X308       0.0082
253  X267       0.0085
235  X249       0.0085
308  X323       0.0086
326  X341       0.0089
10    X18       0.0089
329  X344       0.0089
278  X292       0.0091
35    X44       0.0091
288  X302       0.0091
89    X99       0.0092
161  X173       0.0099
260  X274       0.0105
277  X291       0.0105
364  X379       0.0106
23    X32       0.0107
290  X305       0.0107
182  X196       0.0112
32    X41       0.0113
67    X77       0.0117
181  X195       0.0117
38    X47       0.0119
54    X63       0.0121
96   X106       0.0126
356  X371       0.0127
98   X108       0.0130
147  X159       0.0139
189  X203       0.0142
48    X57       0.0143
130  X141       0.0151
273  X287       0.0154
164  X176       0.0156
197  X211       0.0158
2     X10       0.0159
217  X231       0.0164
358  X373       0.0166
241  X255       0.0166
173  X185       0.0181
162  

In [8]:
# Now,we will remove the variance 0 and low variance < 0.2
var = variance.loc[variance['values']<0.2 , 'ID']
data1 = data.drop(var,axis = 1)
data1.drop('ID',axis = 1,inplace = True)
data1.head()

y  X0 X1  X2 X3 X4 X5 X6 X8  X14  X27  X46  X51  X58  X64  X85  X100  \
0 130.8100   k  v  at  a  d  u  j  o    0    0    1    0    1    0    1     0   
1  88.5300   k  t  av  e  d  y  l  o    0    1    0    1    0    0    1     1   
2  76.2600  az  w   n  c  d  x  j  x    0    1    1    1    1    0    1     0   
3  80.6200  az  t   n  f  d  x  l  e    0    1    1    0    0    0    0     0   
4  78.0200  az  v   n  f  d  h  d  n    0    1    1    1    0    0    0     0   

   X115  X118  X119  X127  X132  X137  X156  X157  X163  X171  X178  X186  \
0     0     1     1     0     0     1     1     0     0     0     0     0   
1     0     1     1     1     1     0     1     0     0     0     1     0   
2     0     0     0     0     1     1     0     1     1     0     0     0   
3     0     0     0     0     1     0     0     1     0     0     0     0   
4     0     0     0     0     1     0     0     1     0     0     0     0   

   X187  X191  X194  X218  X220  X223  X224  X246  X250  X251  X261  X273  \
0     1     0     1     0     1     0     0     0     0     0     0     1   
1     1     0     1     0     0     0     0     0     1     0     0     1   
2     0     0     1     1     1     1     1     1     1     0     0     1   
3     0     0     1     1     1     1     0     1     1     0     0     1   
4     0     0     1     1     1     1     0     1     1     0     0     1   

   X311  X313  X314  X324  X329  X334  X337  X350  X351  X355  X358  X362  \
0     0     0     0     1     1     1     0     0     0     0     0     0   
1     1     0     0     0     1     0     1     0     0     0     0     0   
2     0     0     0     1     0     1     0     1     0     0     1     0   
3     0     0     0     0     0     0     0     1     0     0     1     0   
4     0     0     0     0     0     1     0     1     0     0     1     0   

   X375  X377    source  
0     0     1  train_df  
1     1     0  train_df  
2     0     0  train_df  
3     0     0  train_df  
4     0     0  train_df

In [9]:
data1.shape

(8418, 56)

#### Check for null and unique values for test and train sets.

In [ ]:
#Missing Values in data
data1[data1.isnull().any(axis=1)]

In [11]:
data1.isnull().sum()

y         4209
X0           0
X1           0
X2           0
X3           0
X4           0
X5           0
X6           0
X8           0
X14          0
X27          0
X46          0
X51          0
X58          0
X64          0
X85          0
X100         0
X115         0
X118         0
X119         0
X127         0
X132         0
X137         0
X156         0
X157         0
X163         0
X171         0
X178         0
X186         0
X187         0
X191         0
X194         0
X218         0
X220         0
X223         0
X224         0
X246         0
X250         0
X251         0
X261         0
X273         0
X311         0
X313         0
X314         0
X324         0
X329         0
X334         0
X337         0
X350         0
X351         0
X355         0
X358         0
X362         0
X375         0
X377         0
source       0
dtype: int64

In [12]:
data1.describe()

y       X14       X27       X46       X51       X58       X64  \
count 4209.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000   
mean   100.6693    0.4280    0.6819    0.5955    0.7182    0.5797    0.3677   
std     12.6794    0.4948    0.4658    0.4908    0.4499    0.4936    0.4822   
min     72.1100    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   
25%     90.8200    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   
50%     99.1500    0.0000    1.0000    1.0000    1.0000    1.0000    0.0000   
75%    109.0100    1.0000    1.0000    1.0000    1.0000    1.0000    1.0000   
max    265.3200    1.0000    1.0000    1.0000    1.0000    1.0000    1.0000   

            X85      X100      X115      X118      X119      X127      X132  \
count 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000   
mean     0.4144    0.6867    0.2897    0.6222    0.6222    0.4906    0.6882   
std      0.4926    0.4638    0.4537    0.4849    0.4849    0.4999    0.4633   
min      0.0000    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   
25%      0.0000    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   
50%      0.0000    1.0000    0.0000    1.0000    1.0000    0.0000    1.0000   
75%      1.0000    1.0000    1.0000    1.0000    1.0000    1.0000    1.0000   
max      1.0000    1.0000    1.0000    1.0000    1.0000    1.0000    1.0000   

           X137      X156      X157      X163      X171      X178      X186  \
count 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000   
mean     0.5861    0.7136    0.2864    0.3089    0.6619    0.5589    0.5340   
std      0.4926    0.4521    0.4521    0.4621    0.4731    0.4965    0.4989   
min      0.0000    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   
25%      0.0000    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   
50%      1.0000    1.0000    0.0000    0.0000    1.0000    1.0000    1.0000   
75%      1.0000    1.0000    1.0000    1.0000    1.0000    1.0000    1.0000   
max      1.0000    1.0000    1.0000    1.0000    1.0000    1.0000    1.0000   

           X187      X191      X194      X218      X220      X223      X224  \
count 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000   
mean     0.4244    0.4735    0.4661    0.3168    0.5593    0.5557    0.3154   
std      0.4943    0.4993    0.4989    0.4653    0.4965    0.4969    0.4647   
min      0.0000    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   
25%      0.0000    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   
50%      0.0000    0.0000    0.0000    0.0000    1.0000    1.0000    0.0000   
75%      1.0000    1.0000    1.0000    1.0000    1.0000    1.0000    1.0000   
max      1.0000    1.0000    1.0000    1.0000    1.0000    1.0000    1.0000   

           X246      X250      X251      X261      X273      X311      X313  \
count 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000   
mean     0.4132    0.5544    0.3940    0.4287    0.7162    0.5980    0.2966   
std      0.4924    0.4971    0.4887    0.4949    0.4509    0.4903    0.4568   
min      0.0000    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   
25%      0.0000    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   
50%      0.0000    1.0000    0.0000    0.0000    1.0000    1.0000    0.0000   
75%      1.0000    1.0000    1.0000    1.0000    1.0000    1.0000    1.0000   
max      1.0000    1.0000    1.0000    1.0000    1.0000    1.0000    1.0000   

           X314      X324      X329      X334      X337      X350      X351  \
count 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000 8418.0000   
mean     0.4400    0.5797    0.4367    0.4571    0.5217    0.3400    0.2922   
std      0.4964    0.4936    0.4960    0.4982    0.4996    0.4737    0.4548   
min      0.0000    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   
25%      0.0000    0.0000    0.0000    0.0000    0.0000    

In [13]:
#fill the missing values by mean
data1['y'].fillna(data1['y'].mean(), inplace=True)

In [14]:
data1.isnull().sum()

y         0
X0        0
X1        0
X2        0
X3        0
X4        0
X5        0
X6        0
X8        0
X14       0
X27       0
X46       0
X51       0
X58       0
X64       0
X85       0
X100      0
X115      0
X118      0
X119      0
X127      0
X132      0
X137      0
X156      0
X157      0
X163      0
X171      0
X178      0
X186      0
X187      0
X191      0
X194      0
X218      0
X220      0
X223      0
X224      0
X246      0
X250      0
X251      0
X261      0
X273      0
X311      0
X313      0
X314      0
X324      0
X329      0
X334      0
X337      0
X350      0
X351      0
X355      0
X358      0
X362      0
X375      0
X377      0
source    0
dtype: int64

###Now, there is no missing values in any columns

In [15]:
# Check for unique values in the data set
print(data1.nunique())

y         2546
X0          53
X1          27
X2          50
X3           7
X4           4
X5          33
X6          12
X8          25
X14          2
X27          2
X46          2
X51          2
X58          2
X64          2
X85          2
X100         2
X115         2
X118         2
X119         2
X127         2
X132         2
X137         2
X156         2
X157         2
X163         2
X171         2
X178         2
X186         2
X187         2
X191         2
X194         2
X218         2
X220         2
X223         2
X224         2
X246         2
X250         2
X251         2
X261         2
X273         2
X311         2
X313         2
X314         2
X324         2
X329         2
X334         2
X337         2
X350         2
X351         2
X355         2
X358         2
X362         2
X375         2
X377         2
source       2
dtype: int64


In [16]:
data1.dtypes

y         float64
X0         object
X1         object
X2         object
X3         object
X4         object
X5         object
X6         object
X8         object
X14         int64
X27         int64
X46         int64
X51         int64
X58         int64
X64         int64
X85         int64
X100        int64
X115        int64
X118        int64
X119        int64
X127        int64
X132        int64
X137        int64
X156        int64
X157        int64
X163        int64
X171        int64
X178        int64
X186        int64
X187        int64
X191        int64
X194        int64
X218        int64
X220        int64
X223        int64
X224        int64
X246        int64
X250        int64
X251        int64
X261        int64
X273        int64
X311        int64
X313        int64
X314        int64
X324        int64
X329        int64
X334        int64
X337        int64
X350        int64
X351        int64
X355        int64
X358        int64
X362        int64
X375        int64
X377        int64
source    

In [17]:
#Code all categorical variables as numeric using 'label Encoder' from sklearn's ptreprocessing module
#Initialize the Label Encoder
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
cat_cols = data1.select_dtypes(include=['object']).columns.tolist()
for col in cat_cols:
    print(f"Unique values in {col} before encoding: {data1[col].unique()}")
    data1[col] = le.fit_transform(data1[col].astype(str))
    print(f"Unique values in {col} after encoding: {data1[col].unique()}")


Unique values in X0 before encoding: ['k' 'az' 't' 'al' 'o' 'w' 'j' 'h' 's' 'n' 'ay' 'f' 'x' 'y' 'aj' 'ak' 'am'
 'z' 'q' 'at' 'ap' 'v' 'af' 'a' 'e' 'ai' 'd' 'aq' 'c' 'aa' 'ba' 'as' 'i'
 'r' 'b' 'ax' 'bc' 'u' 'ad' 'au' 'm' 'l' 'aw' 'ao' 'ac' 'g' 'ab' 'av' 'ag'
 'an' 'ae' 'p' 'bb']
Unique values in X0 after encoding: [37 24 46 11 41 49 36 34 45 40 23 32 50 51  9 10 12 52 43 18 15 48  6  0
 31  8 30 16 29  1 26 17 35 44 25 22 28 47  4 19 39 38 21 14  3 33  2 20
  7 13  5 42 27]
Unique values in X1 before encoding: ['v' 't' 'w' 'b' 'r' 'l' 's' 'aa' 'c' 'a' 'e' 'h' 'z' 'j' 'o' 'u' 'p' 'n'
 'i' 'y' 'd' 'f' 'm' 'k' 'g' 'q' 'ab']
Unique values in X1 after encoding: [23 21 24  3 19 13 20  1  4  0  6  9 26 11 16 22 17 15 10 25  5  7 14 12
  8 18  2]
Unique values in X2 before encoding: ['at' 'av' 'n' 'e' 'as' 'aq' 'r' 'ai' 'ak' 'm' 'a' 'k' 'ae' 's' 'f' 'd'
 'ag' 'ay' 'ac' 'ap' 'g' 'i' 'aw' 'y' 'b' 'ao' 'al' 'h' 'x' 'au' 't' 'an'
 'z' 'ah' 'p' 'am' 'j' 'q' 'af' 'l' 'aa' 'c' 'o' 'ar' 'aj' 'ax' 'ab

In [18]:
# Check for unique values in the data set
print(data1.nunique())

y         2546
X0          53
X1          27
X2          50
X3           7
X4           4
X5          33
X6          12
X8          25
X14          2
X27          2
X46          2
X51          2
X58          2
X64          2
X85          2
X100         2
X115         2
X118         2
X119         2
X127         2
X132         2
X137         2
X156         2
X157         2
X163         2
X171         2
X178         2
X186         2
X187         2
X191         2
X194         2
X218         2
X220         2
X223         2
X224         2
X246         2
X250         2
X251         2
X261         2
X273         2
X311         2
X313         2
X314         2
X324         2
X329         2
X334         2
X337         2
X350         2
X351         2
X355         2
X358         2
X362         2
X375         2
X377         2
source       2
dtype: int64


##### There are several techniques for dimensionality reduction, but one common technique is Principal Component Analysis (PCA). PCA is a technique that transforms a set of correlated variables into a smaller set of uncorrelated variables called principal components.

In [19]:
## Perform PCA on the numerical columns
from sklearn.decomposition import PCA
num_cols = data1.select_dtypes(include = [np.number]).columns.tolist()
pca = PCA(n_components = 5)
data_pca = pd.DataFrame(pca.fit_transform(data1[num_cols]))
data_pca.columns = ['PC1','PC2','PC3','PC4','PC5']
data_pca.index = data1.index

In [20]:
# Combine the PCA components with the categorical columns
data2 = pd.concat([data_pca, data1[cat_cols]], axis=1)

In [21]:
# Check for unique values in the data set
print(data2.nunique())

PC1       8099
PC2       8099
PC3       8099
PC4       8099
PC5       8099
X0          53
X1          27
X2          50
X3           7
X4           4
X5          33
X6          12
X8          25
source       2
dtype: int64


In [22]:
# Print explained variance ratio of the first two principal components
print(pca.explained_variance_ratio_[:5])

[0.37942288 0.18846195 0.12257935 0.11310477 0.09293651]


In [23]:
# Get the explained variance ratio of all components
variance_ratio = pca.explained_variance_ratio_
variance_ratio

array([0.37942288, 0.18846195, 0.12257935, 0.11310477, 0.09293651])

In [24]:
# Get the cumulative explained variance ratio of all components
cumulative_variance_ratio = np.cumsum(variance_ratio)
cumulative_variance_ratio

array([0.37942288, 0.56788483, 0.69046418, 0.80356895, 0.89650546])

In [74]:
data1.head()

y  X0  X1  X2  X3  X4  X5  X6  X8  X14  X27  X46  X51  X58  X64  X85  \
0 130.8100  37  23  20   0   3  27   9  14    0    0    1    0    1    0    1   
1  88.5300  37  21  22   4   3  31  11  14    0    1    0    1    0    0    1   
2  76.2600  24  24  38   2   3  30   9  23    0    1    1    1    1    0    1   
3  80.6200  24  21  38   5   3  30  11   4    0    1    1    0    0    0    0   
4  78.0200  24  23  38   5   3  14   3  13    0    1    1    1    0    0    0   

   X100  X115  X118  X119  X127  X132  X137  X156  X157  X163  X171  X178  \
0     0     0     1     1     0     0     1     1     0     0     0     0   
1     1     0     1     1     1     1     0     1     0     0     0     1   
2     0     0     0     0     0     1     1     0     1     1     0     0   
3     0     0     0     0     0     1     0     0     1     0     0     0   
4     0     0     0     0     0     1     0     0     1     0     0     0   

   X186  X187  X191  X194  X218  X220  X223  X224  X246  X250  X251  X261  \
0     0     1     0     1     0     1     0     0     0     0     0     0   
1     0     1     0     1     0     0     0     0     0     1     0     0   
2     0     0     0     1     1     1     1     1     1     1     0     0   
3     0     0     0     1     1     1     1     0     1     1     0     0   
4     0     0     0     1     1     1     1     0     1     1     0     0   

   X273  X311  X313  X314  X324  X329  X334  X337  X350  X351  X355  X358  \
0     1     0     0     0     1     1     1     0     0     0     0     0   
1     1     1     0     0     0     1     0     1     0     0     0     0   
2     1     0     0     0     1     0     1     0     1     0     0     1   
3     1     0     0     0     0     0     0     0     1     0     0     1   
4     1     0     0     0     0     0     1     0     1     0     0     1   

   X362  X375  X377  source  
0     0     0     1       1  
1     0     1     0       1  
2     0     0     0       1  
3     0     0     0       1  
4     0     0     0       1

We can see that the cumulative explained variance ratio reaches a certain threshold i.e. 89.65% ,the principal components are capturing a sufficient amount of the variance in the data and that further dimensionality reduction may not be necessary.

##### Predict your data using XGBoost

In [75]:
#segregating ind and dep var
X = data1.drop("y",axis=1)
y = data1.loc[:,"y"]
# Splitting
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X, y,test_size = 0.3, random_state = 123)
# Instantiation
import xgboost as xg
xgb_r = xg.XGBRegressor(objective ='reg:squarederror',n_estimators = 10, seed = 123)
# Fitting the model
xgb_r.fit(X_train,y_train)
# Predict the model
y_pred = xgb_r.predict(X_test)
###### mape
d = pd.DataFrame()
d["y_test"] = y_test
d["y_pred"] = y_pred
d["mp"] = abs((d["y_test"]- d["y_pred"])/d["y_test"])
(d.mp.mean())*100#mape

3.6994544380775864

In [86]:
# Evaluate the performance of the model on the test data
from sklearn.metrics import mean_squared_error
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
print("RMSE:", rmse)

RMSE: 6.6585000287749505


hence, the accuracy of this model is about 94% which is very good.